### ISSUE TITLE
custom loss function only receives one/first output from model declaring multiple outputs

### ISSUE THREAD

**Issue by r25hbgh** [tensorflow/tensorflow#83168]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

2.18.0

### Custom code

Yes

### OS platform and distribution

CentOS8_64

### Python version

3.9.20

### GCC/compiler version

11.5.0

### Current behavior?

Declare a model with 2 named outputs, one for the bounding box and the second for the label classification.  
Implicit declaration of loss functions in the compile configuration works ok.
Explicit declaration of a custom_loss function in the compile configuration calls the custom_loss function, but it only receives the first output (bbox), not the expected two (bbox + labels)

### Standalone code to reproduce the issue

```shell
Model as (num_classes = 6 for ex):
bbox= layers.Dense(4, name="bbox")(features)
classification_output = layers.Dense(num_classes, name="classification", activation="softmax")(features)

model = keras.Model(inputs=inputs, outputs=[bbox, classification_output], name='vit_object_detector_with_class')

# Dictionary for bounding box loss
bbox_loss_dict = {
    "mse_loss": tf.keras.losses.MeanSquaredError(),  # Mean Squared Error (for bounding box regression)
    "mae_loss": tf.keras.losses.MeanAbsoluteError()   # Mean Absolute Error (alternative for bounding boxes)
}

# Dictionary for classification loss
class_loss_dict = {
    "sparse_categorical_crossentropy": tf.keras.losses.SparseCategoricalCrossentropy(from_logits=False),  # Cross-entropy loss for multi-class classification
    "categorical_crossentropy": tf.keras.losses.CategoricalCrossentropy(from_logits=False)  # Another option for multi-class classification if using one-hot encoded labels
}


Implicit declaration of configuration, inclusive the losses that works ok:
    model.compile(
        optimizer='adam',  # or any optimizer
        loss={
            "bbox": bbox_loss_dict["mse_loss"], 
            "classification": class_loss_dict["sparse_categorical_crossentropy"] 
        },
        loss_weights={
        "bbox": 1.0,
        "classification": 1.5
        },
        metrics={
            "bbox": ["mse", "mae"],
            "classification": ["accuracy"]
        }
    )

#Training:

targets = {
        "bbox": bbox_target,  # shape, for ex: (640,4)
        "classification": class_target   # shape for ex: (640,)
    }
model.fit(x_train, targets, epochs=10, batch_size=32)

#That training works correctly.
#Now, if declaring explicitly a custom_loss function:

def custom_loss(y_true, y_pred):

    bbox_true = y_true[0]  # Bounding boxes ground truth
    class_true = y_true[1]  # Class labels ground truth
    
    bbox_pred = y_pred[0]  # Bounding box predictions
    class_pred = y_pred[1]  # Class predictions
...... etc

and declaring the configuration as:
 model.compile(optimizer='adam', loss=custom_loss, metrics=["accuracy"])

the custom_loss function only receives the bbox data (y_pred.shape = (32,4)) but not the label classification. It should be something like: y_pred.shape = [(32,4),(32,)]
```

**Comment by r25hbgh** [comment]
Very probably, the implicit loss definition is training without errors, but only the bbox data, not including the label classificator.
It seems as at the end of the training the   "model.predict(input)" delivers a 2D array containing only the bbox, without label class.

**Comment by barbart** [comment]
Hi @r25hbgh,

It looks like you're experiencing a problem where the custom loss function receives only the first output (bbox) and not both (bbox and classification). Here are some insights and suggestions that might help resolve this:

1. How Keras Handles Outputs and Custom Loss Functions
In Keras, when using multiple outputs, the loss function must handle each output explicitly. For custom loss functions, the inputs y_true and y_pred are typically single arrays corresponding to each output. When a single custom loss function is used for multiple outputs, Keras does not automatically package all outputs into y_pred.

2. Verify Input Shapes to the Loss Function
To ensure both outputs are available, you can define your custom loss function to accept multiple outputs explicitly. Here’s an example:

python

def custom_loss(y_true, y_pred):
    # Ensure y_true and y_pred are tuples/lists containing multiple outputs
    bbox_true, class_true = y_true["bbox"], y_true["classification"]
    bbox_pred, class_pred = y_pred["bbox"], y_pred["classification"]
    
    # Compute losses for each output
    bbox_loss = tf.keras.losses.MeanSquaredError()(bbox_true, bbox_pred)
    class_loss = tf.keras.losses.SparseCategoricalCrossentropy()(class_true, class_pred)
    
    # Combine the losses with weights
    total_loss = bbox_loss + 1.5 * class_loss
    return total_loss
3. Update Model Compilation
Make sure the loss dictionary aligns with the expected input structure for the custom loss function:

python

model.compile(
    optimizer='adam', 
    loss=custom_loss,
    metrics={
        "bbox": ["mse", "mae"],
        "classification": ["accuracy"]
    }
)
4. Alternative Approach
If you want the y_true and y_pred to remain combined arrays instead of dictionaries, you may need to preprocess them inside the custom loss function:

python

def custom_loss(y_true, y_pred):
    bbox_true, class_true = y_true[..., :4], y_true[..., 4:]
    bbox_pred, class_pred = y_pred[..., :4], y_pred[..., 4:]
    
    bbox_loss = tf.keras.losses.MeanSquaredError()(bbox_true, bbox_pred)
    class_loss = tf.keras.losses.SparseCategoricalCrossentropy()(class_true, class_pred)
    
    return bbox_loss + 1.5 * class_loss
5. Debugging Steps
Print Shapes: Add print() statements in the custom loss function to verify the shapes of y_true and y_pred.
Inspect Model Outputs: Confirm the model is correctly outputting both bbox and classification by calling model.predict() and checking the result.
Verify Dataset Targets: Ensure targets contains both bbox and classification data when passed to model.fit().
Let me know if these suggestions help or if additional context is needed. I’d be happy to assist further!

Best regards,
Barbara

**Comment by tilakrayal** [comment]
@r25hbgh,
I ran the code and face a different error, please find the gist [here](https://colab.research.google.com/gist/tilakrayal/b128b12dec6afd501bbf4e5536476900/untitled2282.ipynb) and share all dependencies to replicate the issue or share a colab gist with the reported error. Thank you!

**Comment by anuj8052** [comment]
he issue you're encountering arises because TensorFlow does not automatically pass multiple outputs into a single custom_loss function when multiple outputs are defined for a model. By default, TensorFlow assumes the loss function applies to one specific output.

To handle this correctly, you need to design your custom loss function and model configuration in a way that explicitly handles multiple outputs. Below is a solution to address this:

**Solution: Combine Outputs and Losses in custom_loss**
Use a dictionary-like approach to group outputs and their corresponding losses.
Pass both the bbox and classification outputs and ground truths explicitly into the custom_loss function.

`# Custom loss function for multiple outputs
def custom_loss(y_true, y_pred):
    # Extract ground truth and predictions for bounding boxes
    bbox_true = y_true["bbox"]
    class_true = y_true["classification"]
    
    bbox_pred = y_pred["bbox"]
    class_pred = y_pred["classification"]

    # Calculate individual losses
    bbox_loss = tf.keras.losses.MeanSquaredError()(bbox_true, bbox_pred)
    class_loss = tf.keras.losses.SparseCategoricalCrossentropy()(class_true, class_pred)

    # Combine losses with weights (if necessary)
    total_loss = bbox_loss + 1.5 * class_loss
    return total_loss

# Recompile the model with the custom loss
model.compile(
    optimizer="adam",
    loss=custom_loss,
    metrics={
        "bbox": ["mse", "mae"],
        "classification": ["accuracy"]
    }
)

# Training data targets
targets = {
    "bbox": bbox_target,  # Shape: (640, 4)
    "classification": class_target  # Shape: (640,)
}

# Train the model
model.fit(x_train, targets, epochs=10, batch_size=32)
`
